rust_src/src/frames.rs

```diff
@@ -1,21 +1,110 @@
 //! Generic frame functions.
 
 use remacs_macros::lisp_fn;
-use remacs_sys::{selected_frame as current_frame, Lisp_Frame};
-
+use remacs_sys::{fget_output_method, fget_terminal, selected_frame as current_frame,
+                 Fselect_window, Lisp_Frame, OutputMethod, Qns, Qw32, Qx};
 use lisp::{ExternalPtr, LispObject};
+use objects::eq;
+use std::ptr;
 use lisp::defsubr;
 
 pub type LispFrameRef = ExternalPtr<Lisp_Frame>;
 
+impl LispFrameRef {
+    #[inline]
+    pub fn is_live(&self) -> bool {
+        if unsafe { fget_terminal(self.as_ptr()) == ptr::null() } {
+            false
+        } else {
+            true
+        }
+    }
+
+    #[inline]
+    pub fn set_selected_window(&mut self, window: LispObject) {
+        self.selected_window = window.to_raw();
+    }
+}
+
 /// Return the frame that is now selected.
 #[lisp_fn]
 pub fn selected_frame() -> LispObject {
     unsafe { LispObject::from(current_frame) }
 }
 
+/// Return non-nil if OBJECT is a frame which has not been deleted.
+/// Value is nil if OBJECT is not a live frame.  If object is a live
+/// frame, the return value indicates what sort of terminal device it is
+/// displayed on.  See the documentation of `framep' for possible
+/// return values.
+#[lisp_fn]
+fn frame_live_p(object: LispObject) -> LispObject {
+    if object.as_frame().map_or(false, |f| f.is_live()) {
+        framep(object)
+    } else {
+        LispObject::constant_nil()
+    }
+}
+
+/// Set selected window of FRAME to WINDOW.
+/// FRAME must be a live frame and defaults to the selected one.  If FRAME
+/// is the selected frame, this makes WINDOW the selected window.  Optional
+/// argument NORECORD non-nil means to neither change the order of recently
+/// selected windows nor the buffer list.  WINDOW must denote a live window.
+/// Return WINDOW.
+#[lisp_fn(min = "2")]
+pub fn set_frame_selected_window(
+    frame: LispObject,
+    window: LispObject,
+    norecord: LispObject,
+) -> LispObject {
+    let mut f = frame;
```

Comment:
This pattern is repeated in a few functions in window.c. I was working in the same place this morning and wrote this:

    pub fn frame_live_or_selected(object: LispObject) -> LispFrameRef {
        let frame = if object.is_nil() {
            selected_frame()
        } else {
            object
        }.as_frame_or_error();
    
        if !frame.is_live() {
            wrong_type!(Qframe_live_p, object);
        }
    
        frame
    }
    
    pub fn window_frame_live_or_selected(window: LispObject) -> LispFrameRef {
        if window.is_nil() {
            selected_frame().as_frame_or_error()
        } else {
            window
                .as_valid_window()
                .map_or(frame_live_or_selected(window), |w| {
                    w.frame().as_frame_or_error()
                })
        }
    }

Reply:
LGTM